Source/Particles/ParticleBoundaryBuffer.cpp

```diff
@@ -0,0 +1,200 @@
+#include "Particles/MultiParticleContainer.H"
+
+#include <AMReX_Geometry.H>
+#include <AMReX_ParmParse.H>
+
+ParticleBoundaryBuffer::ParticleBoundaryBuffer ()
+{
+    m_particle_containers.resize(numBoundaries());
+
+    for (int i = 0; i < numBoundaries(); ++i)
+    {
+        m_particle_containers[i].resize(numSpecies());
+    }
+
+    m_do_boundary_buffer.resize(numBoundaries());
+    for (int i = 0; i < numBoundaries(); ++i)
+    {
+        m_do_boundary_buffer[i].resize(numSpecies(), 0);
+    }
+
+    for (int i = 0; i < numBoundaries(); ++i)
+    {
+        for (int ispecies = 0; ispecies < numSpecies(); ++ispecies)
+        {
+            amrex::ParmParse pp_species(getSpeciesNames()[ispecies]);
+            pp_species.query("save_particles_at_xlo", m_do_boundary_buffer[0][ispecies]);
+            pp_species.query("save_particles_at_xhi", m_do_boundary_buffer[1][ispecies]);
+#if AMREX_SPACEDIM == 2
+            pp_species.query("save_particles_at_zlo", m_do_boundary_buffer[2][ispecies]);
+            pp_species.query("save_particles_at_zhi", m_do_boundary_buffer[3][ispecies]);
+#else
+            pp_species.query("save_particles_at_ylo", m_do_boundary_buffer[2][ispecies]);
+            pp_species.query("save_particles_at_yhi", m_do_boundary_buffer[3][ispecies]);
+            pp_species.query("save_particles_at_zlo", m_do_boundary_buffer[4][ispecies]);
+            pp_species.query("save_particles_at_zhi", m_do_boundary_buffer[5][ispecies]);
+#endif
+#ifdef AMREX_USE_EB
+            pp_species.query("save_particles_at_eb", m_do_boundary_buffer[AMREX_SPACEDIM*2][ispecies]);
+#endif
+        }
+    }
+}
+
+void ParticleBoundaryBuffer::printNumParticles () const {
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim)
+    {
+        for (int iside = 0; iside < 2; ++iside)
+        {
+            auto& buffer = m_particle_containers[2*idim+iside];
+            for (int i = 0; i < numSpecies(); ++i)
+            {
+                int np = buffer[i].isDefined() ? buffer[i].TotalNumberOfParticles(false) : 0;
+                amrex::Print() << "Species " << getSpeciesNames()[i] << " has "
+                               << np << " particles in the boundary buffer "
+                               << " for side " << iside << " of dim " << idim << "\n";
+            }
+        }
+    }
+#ifdef AMREX_USE_EB
+    auto& buffer = m_particle_containers[2*AMREX_SPACEDIM];
+    for (int i = 0; i < numSpecies(); ++i)
+    {
+        int np = buffer[i].isDefined() ? buffer[i].TotalNumberOfParticles(false) : 0;
+        amrex::Print() << "Species " << getSpeciesNames()[i] << " has "
+                       << np << " particles in the EB boundary buffer \n";
+    }
+#endif
+}
+
+void ParticleBoundaryBuffer::clearParticles () {
+    for (int i = 0; i < numBoundaries(); ++i)
+    {
+        auto& buffer = m_particle_containers[i];
+        if (buffer[i].isDefined())
+        {
+            buffer[i].clearParticles();
+        }
```

Comment:
This should be something like this, right?
```suggestion
        for (int ispecies = 0; ispecies < numSpecies(); ++ispecies)
        {
            auto& species_buffer = buffer[ispecies];
            if (species_buffer.isDefined()) species_buffer.clearParticles();
        }
```